File: backend/identity/src/test/java/com/commonsengine/identity/service/MembershipServiceTest.java
package com.commonsengine.identity.service;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertFalse;

import com.commonsengine.identity.domain.MemberRole;
import com.commonsengine.identity.domain.MemberStatus;
import com.commonsengine.identity.domain.Model.Member;
import com.commonsengine.identity.domain.Model.MemberId;
import com.commonsengine.identity.domain.Model.WorkerProfile;
import com.commonsengine.platform.domain.ServiceType;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest
@ActiveProfiles("test")
@Transactional
class MembershipServiceTest {

    @Autowired
    private MembershipService service;

    @Test
    void registerCreatesNewMember() {
        Member member = service.register("张三", "13800000001", Set.of(MemberRole.WORKER));

        assertEquals("张三", member.name());
        assertEquals(MemberStatus.ACTIVE, member.status());
        assertTrue(member.roles().contains(MemberRole.WORKER));
    }

    @Test
    void findByIdReturnsRegisteredMember() {
        Member member = service.register("李四", "13800000002", Set.of(MemberRole.CONSUMER));

        Optional<Member> found = service.findById(member.id());
        assertTrue(found.isPresent());
        assertEquals(member.name(), found.get().name());
    }

    @Test
    void findByIdReturnsEmptyForUnknownId() {
        Optional<Member> found = service.findById(new MemberId("nonexistent"));
        assertTrue(found.isEmpty());
    }

    @Test
    void suspendChangesStatusToSuspended() {
        Member member = service.register("王五", "13800000003", Set.of(MemberRole.WORKER));

        Optional<Member> suspended = service.suspend(member.id(), "违反行为准则");
        assertTrue(suspended.isPresent());
        assertEquals(MemberStatus.SUSPENDED, suspended.get().status());
    }

    @Test
    void withdrawChangesStatusToWithdrawn() {
        Member member = service.register("赵六", "13800000004", Set.of(MemberRole.WORKER));

        Optional<Member> withdrawn = service.withdraw(member.id());
        assertTrue(withdrawn.isPresent());
        assertEquals(MemberStatus.WITHDRAWN, withdrawn.get().status());
    }

    @Test
    void roleStatisticsCountsActiveMembersPerRole() {
        service.register("worker1", "1", Set.of(MemberRole.WORKER));
        service.register("worker2", "2", Set.of(MemberRole.WORKER));
        service.register("consumer1", "3", Set.of(MemberRole.CONSUMER));
        service.register("both", "4", Set.of(MemberRole.WORKER, MemberRole.CONSUMER));

        Map<MemberRole, Integer> stats = service.roleStatistics();
        assertEquals(3, stats.get(MemberRole.WORKER));   // worker1, worker2, both
        assertEquals(2, stats.get(MemberRole.CONSUMER)); // consumer1, both
    }

    @Test
    void registerWorkerProfileSucceedsForExistingMember() {
        Member member = service.register("骑手小王", "13900000001", Set.of(MemberRole.WORKER));
        WorkerProfile profile = new WorkerProfile(
                member.id(),
                Set.of(ServiceType.FOOD_DELIVERY, ServiceType.ERRAND),
                "海淀区",
                null,
                0,
                0
        );

        WorkerProfile saved = service.registerWorkerProfile(profile);
        assertEquals(member.id(), saved.memberId());
        assertTrue(saved.serviceTypes().contains(ServiceType.FOOD_DELIVERY));

        // 验证可回查
        Optional<WorkerProfile> found = service.findWorkerProfile(member.id());
        assertTrue(found.isPresent());
        assertEquals("海淀区", found.get().workRegion());
    }

    @Test
    void registerWorkerProfileFailsForNonExistentMember() {
        WorkerProfile profile = new WorkerProfile(
                new MemberId("ghost"),
                Set.of(ServiceType.RIDE_HAILING),
                "朝阳区",
                null,
                0,
                0
        );

        org.junit.jupiter.api.Assertions.assertThrows(
                IllegalArgumentException.class,
                () -> service.registerWorkerProfile(profile));
    }

    @Test
    void registerRejectsBlankName() {
        org.junit.jupiter.api.Assertions.assertThrows(
                IllegalArgumentException.class,
                () -> service.register("", "123", Set.of(MemberRole.WORKER)));
    }

    @Test
    void registerRejectsEmptyRoles() {
        org.junit.jupiter.api.Assertions.assertThrows(
                IllegalArgumentException.class,
                () -> service.register("test", "123", Set.of()));
    }
}

File: backend/identity/src/test/kotlin/com/commonsengine/identity/service/MembershipServiceTest.kt
package com.commonsengine.identity.service

import com.commonsengine.identity.domain.MemberId
import com.commonsengine.identity.domain.MemberRole
import com.commonsengine.identity.domain.MemberStatus
import com.commonsengine.platform.domain.ServiceType
import com.commonsengine.identity.domain.WorkerProfile
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@ActiveProfiles("test")
@Transactional
class MembershipServiceTest {

    @Autowired
    private lateinit var service: MembershipService

    @Test
    fun `register creates a new member`() {
        val member = service.register("张三", "13800000001", setOf(MemberRole.WORKER))

        assertEquals("张三", member.name)
        assertEquals(MemberStatus.ACTIVE, member.status)
        assertTrue(MemberRole.WORKER in member.roles)
    }

    @Test
    fun `findById returns registered member`() {
        val member = service.register("李四", "13800000002", setOf(MemberRole.CONSUMER))

        val found = service.findById(member.id)
        assertNotNull(found)
        assertEquals(member.name, found!!.name)
    }

    @Test
    fun `findById returns null for unknown id`() {
        val found = service.findById(MemberId("nonexistent"))
        assertNull(found)
    }

    @Test
    fun `suspend changes status to SUSPENDED`() {
        val member = service.register("王五", "13800000003", setOf(MemberRole.WORKER))

        val suspended = service.suspend(member.id, "违反行为准则")
        assertNotNull(suspended)
        assertEquals(MemberStatus.SUSPENDED, suspended!!.status)
    }

    @Test
    fun `withdraw changes status to WITHDRAWN`() {
        val member = service.register("赵六", "13800000004", setOf(MemberRole.WORKER))

        val withdrawn = service.withdraw(member.id)
        assertNotNull(withdrawn)
        assertEquals(MemberStatus.WITHDRAWN, withdrawn!!.status)
    }

    @Test
    fun `roleStatistics counts active members per role`() {
        service.register("worker1", "1", setOf(MemberRole.WORKER))
        service.register("worker2", "2", setOf(MemberRole.WORKER))
        service.register("consumer1", "3", setOf(MemberRole.CONSUMER))
        service.register("both", "4", setOf(MemberRole.WORKER, MemberRole.CONSUMER))

        val stats = service.roleStatistics()
        assertEquals(3, stats[MemberRole.WORKER])   // worker1, worker2, both
        assertEquals(2, stats[MemberRole.CONSUMER])  // consumer1, both
    }

    @Test
    fun `registerWorkerProfile succeeds for existing member`() {
        val member = service.register("骑手小王", "13900000001", setOf(MemberRole.WORKER))
        val profile = WorkerProfile(
            memberId = member.id,
            serviceTypes = setOf(ServiceType.FOOD_DELIVERY, ServiceType.ERRAND),
            workRegion = "海淀区",
        )

        val saved = service.registerWorkerProfile(profile)
        assertEquals(member.id, saved.memberId)
        assertTrue(ServiceType.FOOD_DELIVERY in saved.serviceTypes)

        // 验证可回查
        val found = service.findWorkerProfile(member.id)
        assertNotNull(found)
        assertEquals("海淀区", found!!.workRegion)
    }

    @Test
    fun `registerWorkerProfile fails for non-existent member`() {
        val profile = WorkerProfile(
            memberId = MemberId("ghost"),
            serviceTypes = setOf(ServiceType.RIDE_HAILING),
            workRegion = "朝阳区",
        )

        assertThrows<IllegalArgumentException> {
            service.registerWorkerProfile(profile)
        }
    }

    @Test
    fun `register rejects blank name`() {
        assertThrows<IllegalArgumentException> {
            service.register("", "123", setOf(MemberRole.WORKER))
        }
    }

    @Test
    fun `register rejects empty roles`() {
        assertThrows<IllegalArgumentException> {
            service.register("test", "123", emptySet())
        }
    }
}
